### ISSUE TITLE
"AND" is never appended to SQL predicates

### ISSUE THREAD

**Issue by markostojanovic087** [rs/rest-layer#245]
When I was sending http DELETE requests on nested resource path
(_e.g. localhost:8080/api/users/userid/records/recordid_)
I got response saying that SQL query is not right.

I debug the queries and saw that they were formed like:
_SELECT * FROM record WHERE id LIKE recordid iduser LIKE userid_
The AND word was missing.

I searched through the code and found that AND is never added to predicate.

In order to have a workaround I made a nasty hack in [routing.go](https://github.com/rs/rest-layer/blob/master/rest/routing.go)
```
// Query builds a query object from the matched route
func (r *RouteMatch) Query() (*query.Query, *Error) {

	qp := queryParser{rsc: r.Resource()}
	if qp.rsc == nil {
		return nil, &Error{500, "missing resource", nil}
	}

	//TODO : HACK
	doit := false
	// Append route fields to the query
	for _, rp := range r.ResourcePath {
		if rp.Value != nil {
			log.Print("routing APENDUJEM EQUAL")
			if doit {
				qp.q.Predicate = append(qp.q.Predicate, &query.And{})
			} else {
				doit = true
			}

			qp.q.Predicate = append(qp.q.Predicate, &query.Equal{Field: rp.Field, Value: rp.Value})
		}
	}
```

In order to form a sql query properly I also had to change a [query.go](https://github.com/apuigsech/rest-layer-sql/blob/master/query.go) (which is not part of this project), probably due to my lack of understanding of query.Add and other Expressions.

Not sure if it is relevant, but what I did there is
```
func translatePredicate(q query.Predicate) (sqlQuery string, sqlParams []interface{}, err error) {
	log.Print(q)
	for _, exp := range q {
		switch t := exp.(type) {
		case *query.And:
			//TODO : HACK
			log.Print("Expression type AND")
			/*var s string
			for _, subExp := range *t {
				log.Print("count")
				sb, extraSqlParams, err := translatePredicate(query.Predicate{subExp})
				if err != nil {
					return "", []interface{}{}, err
				}
				sqlParams = append(sqlParams, extraSqlParams...)
				s += sb + " AND "
			}

			sqlQuery += "(" + s[:len(s)-5] + ")"*/

			sqlQuery += " AND " //Added this
```

Could you please tell me more about this?
Have you experienced this issue?

Thanks!

**Comment by smyrman** [comment]
I presume this is a bug in https://github.com/apuigsech/rest-layer-sql. Could you please raise it again there?

**Comment by smyrman** [comment]
@apuigsech, FYI.

**Comment by Dragomir-Ivanov** [comment]
@markostojanovic087 You may want to test the same with the `rest-layer-mem`. It is very easy to replicate your use-case. In any case it is perfect opportunity to learn some more about `rest-layer`.

@smyrman I am not very sure about that, the code is:
```
	// Append route fields to the query
	for _, rp := range r.ResourcePath {
		if rp.Value != nil {
			qp.q.Predicate = append(qp.q.Predicate, &query.Equal{Field: rp.Field, Value: rp.Value})
		}
	}
```
If there are multiple route fields, they are not wrapped within top-level `shema.And{}`.
Could be very well bug in `rest-layer`

**Comment by rs** [comment]
Predicates are AND by default. It must be a bug in the sql strorer.

**Comment by smyrman** [comment]
@markostojanovic087, I am closing this issue as the SQL storage backed is completely third party. The issue is, as @rs is pointing out, is that the storer backed you are using is not joining the Expressions in a predicate with AND.

The only distinction between `schema.Predicate` and `schema.And` In rest-later, is that the Mongo backed can reliey on implisit AND in queries.

**Comment by markostojanovic087** [comment]
Hello, I debugged the predicate, and there was no "and" in the expression array. Since there was no "and" in it, the rest-layer-sql code was never executing case *query.And:. This case was executed only after I added the given code for appending "and" in [routing.go](https://github.com/rs/rest-layer/blob/master/rest/routing.go). 

If I understand your point, an "and" should be implicitly added (when having two or more equal's) on rest-layer-sql side. Could you please confirm that? 

How can rest-layer-sql know if it needs to add "and" or "or"? 
If it should add "and" implicity, where do you add "or"? 
In order to cover proper functionality when assuming that rest-layer-sql is adding implicit "and", you must manually add every other type of condition operators.

Thanks!

**Comment by rs** [comment]
That’s right, the root level of the query should be interpreted as a AND list.
